control/grpc.py

```diff
@@ -645,6 +669,56 @@ def get_ns_id_message(self, nsid, uuid):
             ns_id_msg = f"using UUID {uuid} "
         return ns_id_msg
 
+    def set_ana_state(self, request, context=None):
+        return self.execute_grpc_function(self.set_ana_state_safe, request, context)
+
+    def set_ana_state_safe(self, ana_info: pb2.ana_info, context=None):
+        """Sets ana state for this gateway."""
+        self.logger.info(f"Received request to set ana states {ana_info.states}")
+        state = self.gateway_state.local.get_state()
+
+        # Iterate over nqn_ana_states in ana_info
+        for nas in ana_info.states:
+            nqn = nas.nqn
+            prefix = f"{self.gateway_state.local.LISTENER_PREFIX}{nqn}_{self.gateway_name}_"
+            listener_keys = [key for key in state.keys() if key.startswith(prefix)]
+            self.logger.info(f"Iterate over {nqn=} {prefix=} {listener_keys=}")
+            # fill the static gateway dictinary per nqn and grp_id
+            for gs in nas.states:
+                self.ana_map[nqn][gs.grp_id]  = gs.state
+
+            for listener_key in listener_keys:
+                listener = json.loads(state[listener_key])
+                self.logger.info(f"{listener_key=} {listener=}")
+
+                # Iterate over ana_group_state in nqn_ana_states
+                for gs in nas.states:
+                    # Access grp_id and state
+                    grp_id = gs.grp_id
+                    ana_state = "optimized" if gs.state == pb2.ana_state.OPTIMIZED else "inaccessible"
```

Comment:
I understand that these values come from the ANA protocol and they are fixed, but still it's better form to use symbolic constant names and not hardcoded values. Even better, as you have an enum for the state you can make sure the constant names in the enum definition match the actual ANA values and use the  GatewayEnumUtils.get_value_from_key() function to retrieve them. If you prefer to keep the enum names in upper case, just use lower() on the returned value.

Reply:
@gbregman we're mapping the gateway's interface gRPC [ana_state](https://github.com/baum/ceph-nvmeof/blob/0438f770642d85ac2991b0603c46093b726f3c3e/control/proto/gateway.proto#L234-L249), into SPDK JSON RPC, see [nvmf_subsystem_listener_set_ana_state method](https://spdk.io/doc/jsonrpc.html).  Added a comment in the code, let me know if additional changes required.

Reply:
@baum I understand that and I don't mean that you change the actual string values. But, instead of using hardcoded strings in grpc.py you could use the following in proto/gateway.proto:
```
enum ana_state {
  unset         = 0;
  optimized     = 1;
  non_optimized = 2;
  inaccessible  = 3;
}
```
and whenever you need the string you can use:
```
str_value = GatewayEnumUtils.get_value_from_key(pb2.ana_state, int_value)
```

Or, you can leave the enum definition in gateway.proto as it is now and use:
```
GatewayEnumUtils.get_value_from_key(pb2.ana_state, int_value).lower()
```

Reply:
Cool GatewayEnumUtils logic, not sure necessary here if we point to the SPDK JSON format in the comments.